spec/virtual_total_spec.rb

```diff
@@ -435,37 +435,144 @@ def model_with_children(count)
+          expect do
+            expect(query.map(&:avg_recently_published_books_rating)).to eq([3, 5, 0, 0])
+          end.to make_database_queries(:count => 2)
+        end
+
+        it "calculates avg from attribute (and preloaded association)" do
+          authors
+          query = Author.includes(:recently_published_books).select(:id, :avg_recently_published_books_rating).order(:id).load
+          expect do
+            expect(query.map(&:avg_recently_published_books_rating)).to eq([3, 5, nil, nil])
+          end.to_not make_database_queries
+        end
       end
 
-      it "calculates sum from attribute" do
-        authors
-        query = Author.select(:id, :sum_recently_published_books_rating).order(:id).load
-        expect do
-          expect(query.map(&:sum_recently_published_books_rating)).to eq([6, 5, 0, 0])
-        end.to make_database_queries(:count => 2)
+      describe ":maximum" do
+        # NOTE: rails converts the nil to a 0
+        it "calculates max with one off query" do
+          authors
+
+          expect do
+            expect(authors.map(&:highest_recently_published_books_rating)).to eq([4, 5, 0, 0])
+          end.to make_database_queries(:count => 4)
+        end
+
+        it "calculates max from preloaded association" do
+          authors.each { |a| a.recently_published_books.load }
+
+          expect do
+            expect(authors.map(&:highest_recently_published_books_rating)).to eq([4, 5, nil, nil])
+          end.to_not make_database_queries
+        end
+
+        it "calculates max from attribute" do
+          authors
+          query = Author.select(:id, :highest_recently_published_books_rating).order(:id).load
+          expect do
+            expect(query.map(&:highest_recently_published_books_rating)).to eq([4, 5, 0, 0])
+          end.to make_database_queries(:count => 2)
+        end
+
+        it "calculates max from attribute (and preloaded association)" do
+          authors
+          query = Author.includes(:recently_published_books).select(:id, :highest_recently_published_books_rating).order(:id).load
+          expect do
+            expect(query.map(&:highest_recently_published_books_rating)).to eq([4, 5, nil, nil])
+          end.to_not make_database_queries
+        end
       end
 
-      it "calculates sum from attribute (and preloaded association)" do
-        authors
-        query = Author.includes(:recently_published_books).select(:id, :sum_recently_published_books_rating).order(:id).load
-        expect do
-          expect(query.map(&:sum_recently_published_books_rating)).to eq([6, 5, nil, 0])
-        end.to_not make_database_queries
+      describe ":minimum" do
+        # NOTE: rails converts the nil to a 0
+        it "calculates min with one off query" do
+          authors
+
+          expect do
+            expect(authors.map(&:lowest_recently_published_books_rating)).to eq([2, 5, 0, 0])
```

Comment:
this is our standard.

We need to find a way to get the others to return the same value as this.

Reply:
i think we're good on this comment, i think i did mess up the `orders by values with a nil (having the nil (defaulted to 0) first` tests though